### PR TITLE
Change evm_hook to use Transaction Receipt

### PR DESCRIPTION
**Pull request by ramacarlucho**
Closes: #ENG-381

## Description

Some hooks may need gasUsed by the execution
______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/evm/keeper/state_transition.go [hunk 1]
```diff
@@ -220,13 +220,29 @@ func (k *Keeper) ApplyTransaction(tx *ethtypes.Transaction) (*types.MsgEthereumT
...
+		PostState: nil,
+		// NOTE: assume until this point that tx is successful
+		Status: ethtypes.ReceiptStatusSuccessful,
+		Bloom:  ethtypes.BytesToBloom(k.GetBlockBloomTransient().Bytes()),
+		// Cumulative gas used for the current block, excluding this tx
+		CumulativeGasUsed: ctx.BlockGasMeter().GasConsumedToLimit(),
+		TxHash:            txHash,
+		Logs:              logs,
+		GasUsed:           res.GasUsed,
+		ContractAddress:   *tx.To(),
```
Comment (fedekunze) [inline x/evm/keeper/state_transition.go:234]: TODO: fix potential panic

File: x/evm/keeper/state_transition.go [hunk 2]
```diff
@@ -220,13 +220,29 @@ func (k *Keeper) ApplyTransaction(tx *ethtypes.Transaction) (*types.MsgEthereumT
...
+		BlockHash:         common.BytesToHash(ctx.HeaderHash()),
+		BlockNumber:       big.NewInt(ctx.BlockHeight()),
+		TransactionIndex:  uint(k.GetTxIndexTransient()),
```
Comment (fedekunze) [inline x/evm/keeper/state_transition.go:237]: TODO: should be current idx +1

File: x/evm/keeper/state_transition.go (hunk 1 above)
Comment (ramacarlucho) [reply x/evm/keeper/state_transition.go:234]: we dont know if its a `ContractAddress` at this stage.

File: x/evm/keeper/state_transition.go [hunk 3]
```diff
@@ -360,10 +362,11 @@ func (k *Keeper) ApplyMessageWithConfig(msg core.Message, tracer vm.Tracer, comm
 		// - reset sender's nonce to msg.Nonce() before calling evm.
 		// - increase sender's nonce by one no matter the result.
 		k.SetNonce(sender.Address(), msg.Nonce())
-		ret, _, leftoverGas, vmErr = evm.Create(sender, msg.Data(), leftoverGas, msg.Value())
+		ret, contractAddr, leftoverGas, vmErr = evm.Create(sender, msg.Data(), leftoverGas, msg.Value())
 		k.SetNonce(sender.Address(), msg.Nonce()+1)
 	} else {
 		ret, leftoverGas, vmErr = evm.Call(sender, *msg.To(), msg.Data(), leftoverGas, msg.Value())
+		contractAddr = *msg.To()
```
Comment (ramacarlucho) [inline x/evm/keeper/state_transition.go:369]: Shouldnt we check if *msg.To() is a contractAddr?

Comment (fedekunze) [reply x/evm/keeper/state_transition.go:369]: good catch

**Review by danburck — APPROVED** [review]
LGTM

**Review by fedekunze — APPROVED** [review]

File: x/evm/keeper/state_transition.go (hunk 3 above)
Comment (yihuang) [reply x/evm/keeper/state_transition.go]: Isn't the contract address in receipt should just be the contract address created, it's null for contract call?

File: x/evm/keeper/state_transition.go [hunk 4]
```diff
@@ -232,16 +219,53 @@ func (k *Keeper) ApplyTransaction(tx *ethtypes.Transaction) (*types.MsgEthereumT
...
+
+		// Only call hooks if tx executed successfully.
+		if err = k.PostTxProcessing(msg.From(), tx.To(), receipt); err != nil {
+			// If hooks return error, revert the whole tx.
+			res.VmError = types.ErrPostTxProcessing.Error()
+			k.Logger(ctx).Error("tx post processing failed", "error", err)
+		} else if commit != nil {
+			// PostTxProcessing is successful, commit the cache context
+			commit()
+			ctx.EventManager().EmitEvents(k.Ctx().EventManager().Events())
```
Comment (yihuang) [inline x/evm/keeper/state_transition.go:265]: hooks execution is moved under the `k.WithContext(ctx)`, this changed the context hooks executed in.
And `k.Ctx()` is same as `ctx` here, we'll lose events emitted by hooks, and re-emit events in `ctx.

**Pull request merged**